Thumbnails API and Support

tbarsballe opened the pull request:
This pull request constitues a collection of interdependant classes for caching and serving thumbnails of maps and layers:
* ThumbnailController.java provides an API endpoint for retrieving thumbnails and generates thumbnails if they do not exist.
* AppConfiguration.java tracks the location of the cached thumbnail files, and allows this location to be configured similar to the geoserver data directory. (The idea is to add to this class if similar configuration elements are required in the future)
* ComposerOutputFormat.java extends the regular PNG getMap format to add support for saving thumbnails and bounding boxes.
* ThumbnailInvalidatingCatalogListener listens to catalog events and invalidated thumbnail entries to the metadata maps if resources are changed without their thumbnails being updated.

A faitly extensinve integration test touching on all these components has been added to AppIntegrationTest.java

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/AppConfiguration.java:
```diff
@@ -0,0 +1,176 @@
+     * @return String The absolute path to the data directory, or <code>null</code> if it could not
+     * be found. 
+     */
+    private String lookupCacheDirectory(ServletContext servContext) {
```
The cascading property lookup functionality is already provided by `org.geoserver.platform.GeoServerExtensions`

tbarsballe replied:
Thanks. Interesting that the GEOSERVER_DATA_DIR lookup does not use that funtionality (That is where I borrowed parts of this code from).

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/AppConfiguration.java:
```diff
@@ -0,0 +1,176 @@
+        
+        //Use the default of data/composer
+        if(cacheDir == null) {
+            cacheDir = (new GeoServerDataDirectory(catalog.getResourceLoader()).root().getPath())
```
An instance of `GeoServerDataDirectory` should be obtainable via  `GeoServerExtensions.bean(GeoServerDataDirectory.class)`

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/catalog/ThumbnailInvalidatingCatalogListener.java:
```diff
@@ -0,0 +1,187 @@
+     * but the thumbnails for all dependent objects are still invalidated.
+     */
+    @Override
+    public void handleModifyEvent(CatalogModifyEvent event)
```
Is this the same logic as `handleRemoveEvent`?

tbarsballe replied:
Not quite. handleModifyEvent does some extra checking to make sure thumbnail changes are not overwritten by calling updateThumbnailEvent.

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/catalog/ThumbnailInvalidatingCatalogListener.java:
```diff
@@ -0,0 +1,187 @@
+                
+                //Modify thumbnail entry
+                if (oldMap.containsKey(Metadata.THUMBNAIL) && newMap.containsKey(Metadata.THUMBNAIL)
+                        && oldMap.get(Metadata.THUMBNAIL) != newMap.get(Metadata.THUMBNAIL) ) {
```
object equality comparison

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/catalog/ThumbnailInvalidatingCatalogListener.java:
```diff
@@ -0,0 +1,187 @@
+     * If the event modifies the metadata map, and the metadata map contains a thumbnail entry that 
+     * is not changed, invalidates the thumbnail entry by removing it from the metadata map.
+     */
+    private boolean updateThumbnailEvent(CatalogModifyEvent event) {
```
extract to more easily unit testable method using just `oldMap` and `newMap`

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/api/controllers/Metadata.java:
```diff
@@ -18,7 +23,39 @@
+    public static void invalidateThumbnail(PublishedInfo layer) {
+        //Allow a bit of leeway, to support GetMap composer format 
+        //(in case of getMap returning before put layer)
+        Date d = new Date(new Date().getTime()-1000);
```
use `System.currentTimeMillis()` and avoid creating Date objects. Not sure why we're using Date objects anyway (for Metadata) but I guess we're stuck now. `Long` would be preferable.

tbarsballe replied:
I'll update this. A good chunck of the existing API should also probably be refactored to use System.currentTimeMillis() - I remember seeing new Date() in a few places...

ischneider commented on geoserver/webapp/src/test/java/com/boundlessgeo/geoserver/AppIntegrationTest.java:
```diff
@@ -489,4 +501,144 @@ public void testCreateLayerFromRasterResource() throws Exception {
+        getMap.setStyles(Arrays.asList(layer.getDefaultStyle().getStyle()));
+        
+        //file.lastModified is only accurate to the second
+        Thread.sleep(2000);
```
This test now takes 4 seconds no matter what (this sleep call and the one above). Let's figure out how to avoid this.

tbarsballe replied:
To start, I agree that 6 seconds of waiting is not good, even in a test case. However, without a way to get lastModified with better than second precision (this seems to be a Java limitation for Linux systems, and is fixed in Java 8), I'm not sure there is a way to tell if the file has been updated (Especially since the file contents should be identical).
The purpose of this check is to ensure the file actually gets updated when it should, and I feel like this is something that should definately be tested.

Possible fixes include:
* Update when we switch to Java 8. Disable the lastModified check in the meantime?
* Modify the map style so the image returned actually is different, then compare the returned images. This seems hacky and unreliable, especially since two images can be visually identical but actually quite different (risk of false positives)

Do you have any other suggestions?

tbarsballe replied:
I've done a few tests to optimize this value, and 1000ms seems safe enough. Any lower will cause test failures though (800 ms usually fails, 900ms sometimes fails, etc). Without a better way of checking if the file is modified, I think we are stuck with a 3 second delay.

ischneider commented on geoserver/webapp/src/test/java/com/boundlessgeo/geoserver/AppIntegrationTest.java:
```diff
@@ -489,4 +501,144 @@ public void testCreateLayerFromRasterResource() throws Exception {
+        request.setMethod("put");
+        
+        //Delay so invalidation margin is exceeded
+        Thread.sleep(2000);
```
Oops, 6 seconds. See above comment.

ischneider commented on geoserver/webapp/src/test/java/com/boundlessgeo/geoserver/AppIntegrationTest.java:
```diff
@@ -489,4 +501,144 @@ public void testCreateLayerFromRasterResource() throws Exception {
+        GeoServerResourceLoader rl = catalog.getResourceLoader();
+
+        LayerInfo layer = catalog.getLayerByName("sf:PrimitiveGeoFeature");
+        StyleInfo style = catalog.getFactory().createStyle();
```
Possible to use `org.geoserver.importer.StyleGenerator` to avoid this (and the additional test resource) ?

ischneider commented:
Beyond the comments above, nice work.

tbarsballe commented:
Thanks for the review. Also thanks for pointing out those various GeoServer library funcitions. It can sometimes be difficult to determine if there is an existing library do do something, especially when the existing code doesn't use it consistently.
I'll make the changes noted above, although I still don't see how we can reliably test if the file has been updated without the delay.

tbarsballe commented:
Added a bunch of fixes based on comments above.
Additional fixes based on testing by Eva:
* If layerGroup.getStyles().get(i) is null, use default Style for layer i
* Thumbnail size = 75

ischneider commented:
> Thanks for the review.

Happy to - thanks for being patient :smile:  Still wondering about `ComposerOutputFormat` as it seems like it could be reduced to what is actually used (unless I'm missing something...)

> Also thanks for pointing out those various GeoServer library funcitions. It can sometimes be difficult to determine if there is an existing library do do something, especially when the existing code doesn't use it consistently.

Indeed.

> I'll make the changes noted above, although I still don't see how we can reliably test if the file has been updated without the delay.

How about deleting the thumbnail after verifying it has been created correctly? Assuming it _should_ be recreated, this seems like an adequate pre-test condition.

tbarsballe commented:
> Still wondering about ComposerOutputFormat as it seems like it could be reduced to what is actually used (unless I'm missing something...)

Sorry, I'm not sure I saw a comment about the ComposerOutputFormat? The idea is we use this outputFormat for all composer getMap calls (instead of image/png). It automatically generates thumbnails for us and updates other metadata (such as latest view bbox). Then, when we go to get the thumbnail, we can use the last generated one instead of having to do a getMap request to recreate the thumbnail

>> I'll make the changes noted above, although I still don't see how we can reliably test if the file has been updated without the delay.

>How about deleting the thumbnail after verifying it has been created correctly? Assuming it _should_ be recreated, this seems like an adequate pre-test condition.

That fails to test if it is capable of updating the thumbnail when the file already exists in the cache directory, which is more or less the whole point. We already verify it can create a file on an empy cache with the first test.

ischneider commented:
> Still wondering about ComposerOutputFormat as it seems like it could be reduced to what is actually used (unless I'm missing something...)
> 
> Sorry, I'm not sure I saw a comment about the ComposerOutputFormat? The idea is we use this outputFormat for all composer getMap calls (instead of image/png). It automatically generates thumbnails for us and updates other metadata (such as latest view bbox). Then, when we go to get the thumbnail, we can use the last generated one instead of having to do a getMap request to recreate the thumbnail

I guess I'm not seeing how it's actually used. I see that the controller is making the GetMap call and writing out some files there. I also see that the format is used in a test but the controller uses the format 'image/png'.

ischneider commented:
> I'll make the changes noted above, although I still don't see how we can reliably test if the file has been updated without the delay.
> 
>     How about deleting the thumbnail after verifying it has been created correctly? Assuming it should be recreated, this seems like an adequate pre-test condition.
> 
> That fails to test if it is capable of updating the thumbnail when the file already exists in the cache directory, which is more or less the whole point. We already verify it can create a file on an empy cache with the first test.

True. Use `File.setLastModified(long time)` in the test :smile:

What are the chances of concurrent modification happening for thumbnails? I could see this actually being an issue and either resulting in: corrupted thumbnails or file-locking issues. Seems like a coarse-grained lock would suffice. If/when my catalog lock patch gets incorporated, this could be leveraged as well.

tbarsballe commented:
>What are the chances of concurrent modification happening for thumbnails? I could see this actually being an issue and either resulting in: corrupted thumbnails or file-locking issues. Seems like a coarse-grained lock would suffice. If/when my catalog lock patch gets incorporated, this could be leveraged as well.

I feel like we would only get concurrent modification if we are in a multi-user situation. Definately seems like something that should be looked at. I'll see about adding proper locking & checking to the thumbnail service. I think the `Files` library in geoserver has a method designed to create a temp file and copy it into place at the last minute, which is used to avoid concurrent modification in styles.

ischneider commented:
Eva explained the idea with the output format, sorry for not understanding this earlier. I'm thinking this is not needed and violates the principle of doing 1 thing and adds unnecessary complexity.

As currently implemented, it doesn't look suitable for it's use as it creates thumbnails with each request.

tbarsballe commented:
Based on our discussion this afternoon, I have removed the ComposerOutputFormat and migrated several relevant methods to ThumbnailController.

I think this is good to merge now.

tbarsballe commented:
Added basic locking before writing out thumbnails. If you try to write to a thumbnail currently being written, the API request that triggered that write will throw an exception.

At this point, I think I have addressed all the feedback for this PR. Eva seems to have gotten all the frontend work done, so I think we are ready to merge.

Any final feedback @ischneider ?

ischneider commented:
As noted, the FileLock approach does not work properly. To avoid the current issue, instead lock a single dummy file (e.g. `config.getCacheFile(thumbnailPath) + ".lck"`).

I don't like the idea of having the endpoint throw an exception but I'm not sure of the best approach here.

One approach would be to serialize access: e.g. make a subsequent request wait until a preceding one completes. This is the easiest to implement and preserves the desired behavior - the thumbnail represents the state of the most recent request.

Ideally we could 'interrupt' the current request if a subsequent one occurs but I don't believe it's possible to do this. It may be possible to ignore the results of the render (and simply not make the thumbnails, avoiding some time/effort).

The potential danger is having requests back up leading to thread exhaustion.

Finally, thinking about the concurrency issues in general, since it's quite possible to envision a scenario where >5 thumbs are simultaneously being rendered, perhaps we should cap the # of concurrent thumb requests across layers. Otherwise it would be easy to cause inadvertent DOS.

ischneider commented on geoserver/webapp/src/main/java/com/boundlessgeo/geoserver/api/controllers/ThumbnailController.java:
```diff
@@ -0,0 +1,346 @@
+            hiResLock = hiRes.getChannel().lock();
+            loRes.write(scaleImage(os.toByteArray(), 0.5, true));
+            //Don't scale, but crop to square
+            hiRes.write(scaleImage(os.toByteArray(), 1.0, true));
```
Why are we serializing to bytes then re-reading twice? Seems like `scaleImage` should just take a `RenderedImage` instead of the `byte[]`

tbarsballe commented:
>As noted, the FileLock approach does not work properly. To avoid the current issue, instead lock a single dummy file (e.g. config.getCacheFile(thumbnailPath) + ".lck").

This would only work for accesses by the same API on the same geoserver instance. I have done some more research, and have found a proper solution to locking the thumbnail files:
```
            loResRAF = new RandomAccessFile(config.createCacheFile(thumbnailFilename(layer)), "rw");
            loResLock = loResRAF.getChannel().tryLock();
            ...
            loRes = new FileOutputStream(loResRAF.getFD());

```

> I don't like the idea of having the endpoint throw an exception but I'm not sure of the best approach here.

This has been the standard way of telling the UI an error occured throughout the composer api. The Spring MVC translates the exception into a http 500 error, and the front-end translates this into a usefull error message.

> One approach would be to serialize access: e.g. make a subsequent request wait until a preceding one completes. This is the easiest to implement and preserves the desired behavior - the thumbnail represents the state of the most recent request.

This seems like an idea, but we should only serialize on a per-layer basis; see below. Although I feel like this could be excessive for the actuall use case - I am not sure how often we are actually going to update the same thumbnail concurrently (seems like it shouldn't happen very often)

> Ideally we could 'interrupt' the current request if a subsequent one occurs but I don't believe it's possible to do this. It may be possible to ignore the results of the render (and simply not make the thumbnails, avoiding some time/effort).

I also feel like interrupting the request seems like it would not be possible to do. The getMap is easily the most computationally intensive section of the request, so I feel like the effort involved in checking if another request has been sent in the meantime would outweigh any savings from cancelling the rest of the operation.

> The potential danger is having requests back up leading to thread exhaustion.

> Finally, thinking about the concurrency issues in general, since it's quite possible to envision a scenario where >5 thumbs are simultaneously being rendered, perhaps we should cap the # of concurrent thumb requests across layers. Otherwise it would be easy to cause inadvertent DOS.

Most of the effort of thumbnail requests is the getMap request, which is generally publicly accessible, so any request capping should be a the getMap level I would expect. Also, just loading the layers page will trigger up to 25 different thumbnail requests during initial cache generation.

ischneider commented:
The lock change looks like it should work - nice find. Would be good to add the verification to the test.

As for the exception/response during concurrent access, my point was less about the means (throwing an exception) than the result - spurious non-errors that appear as errors are confusing. I don't believe the front-end can translate this specific exception into something useful for the user.

I agree the same layer thumbnail will rarely encounter this issue but it would be good to understand what occurs when it happens.

I understand the issue w/ GetMap and publicly available layers - there are ways to address that issue. And correct, I'm looking to avoid 25 concurrent thumbnail requests.

tbarsballe commented:
> As for the exception/response during concurrent access, my point was less about the means (throwing an exception) than the result - spurious non-errors that appear as errors are confusing.

If the thumbnail is locked, then that means that generating an up-to date thumbnail has failed - This feels like an error to me. 

> I don't believe the front-end can translate this specific exception into something useful for the user.

No. In this specific case, I would expect that the front-end just should just re-try getting the thumbnail. (The front-end mostly translates exceptions based on the context of the request. So you are correct in that within the context of a single request it does not really distinguish between exceptions.)

> And correct, I'm looking to avoid 25 concurrent thumbnail requests.

Note that the 25 requests will only ever happen once - when the cache is first generated (ie. when the Maps/Layers page is first loaded with a clean thumbnail cache). All other times we will likely have 0 or 1 calls upon generating the page.

ischneider commented:
> If the thumbnail is locked, then that means that generating an up-to date thumbnail has failed - This feels like an error to me.

It's not an error, it's concurrent use.

> No. In this specific case, I would expect that the front-end just should just re-try getting the thumbnail. (The front-end mostly translates exceptions based on the context of the request. So you are correct in that within the context of a single request it does not really distinguish between exceptions.)

The 'client' is most likely an img element. It will not re-try getting the thumbnail without some bigger efforts like listening to the image load state, etc. Easier to fix on the backend.

> Note that the 25 requests will only ever happen once - when the cache is first generated (ie. when the Maps/Layers page is first loaded with a clean thumbnail cache). All other times we will likely have 0 or 1 calls upon generating the page.

Generally. If a layer is shared amongst maps, this could trigger a change for all maps. Likewise with a shared style. I agree that _typical_ use will not result in these cases.

tbarsballe commented:
byte[] serialization removed from image processing steps

`Striped.semaphore(4,1)` used to surround the contents of `createThumbnail()`, ensuring at most 4, and using the `PublishedInfo layer` as a key to ensure no concurrent requests against the same layer or layer group. 
If the file is locked when access is attempted, we (still) throw an exception, since this means the file is in use by an external program (The semaphore prevents concurrent use by the API).

tbarsballe commented:
Squashed to single commit for merging. Full commit history available at https://github.com/tbarsballe/suite/tree/thumbnails-expanded
